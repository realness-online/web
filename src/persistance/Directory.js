/** @typedef {import('@/types').Id} Id */
/** @typedef {import('@/types').Type} Type */
/** @typedef {import('@/types').Created} Created */

import { as_path_parts, as_created_at } from '@/utils/itemid'
import { get, set, keys } from 'idb-keyval'
import { directory } from '@/utils/serverless'

/**
 * @implements {Directory}
 */
class Directory {
  id = /** @type {Id} */ ('')

  /**
   * @type {Type[]}
   */
  types = [] // at the root of the directory folders are types in our vocabulary

  /**
   * @type {(Created|Directory)[]}
   */
  archive = []

  /**
   * @type {Id[]}
   */
  items = []

  /**
   * @param {Id} id
   */
  constructor(id) {
    this.id = /** @type {Id} */ (as_directory_id(id))
  }
}

/**
 * @param {unknown} maybe
 * @returns {maybe is Directory}
 */
export const is_directory = maybe =>
  Boolean(maybe) &&
  typeof maybe === 'object' &&
  'id' in maybe &&
  'types' in maybe &&
  'archive' in maybe &&
  'items' in maybe
//

/**
 * @param {Id} itemid
 * @returns {string}
 */
export const as_directory_id = itemid => {
  const [author, type] = as_path_parts(itemid)
  if (itemid?.endsWith('/')) return itemid
  return `/${author}/${type}/index/` // Trailing slash indicates directory
}

/**
 * @param {Id} itemid
 * @returns {Promise<Directory | null>}
 */
export const build_local_directory = async itemid => {
  console.time('build_local_directory')
  const path = /** @type {Id} */ (as_directory_id(itemid))
  const directory = new Directory(path)
  const everything = await keys()
  everything.forEach(itemid => {
    if (as_directory_id(/** @type {Id} */ (itemid)) === path) {
      const id = as_created_at(/** @type {Id} */ (itemid))
      if (id) directory.items.push(/** @type {Id} */ (itemid))
    }
  })
  console.timeEnd('build_local_directory')
  return directory
}

/**
 * @param {Id} itemid
 * @returns {Promise<Directory | null>}
 */
export const load_directory_from_network = async itemid => {
  if (navigator.onLine) {
    const [author, type, created = null, archive = null] = as_path_parts(itemid)

    const path = as_directory_id(itemid)

    const meta = new Directory(/** @type {Id} */ (path))
    console.group('request:directory')
    console.info('itemid', itemid)
    console.info('author', author)
    console.info('type', type)
    console.info('created', created)
    console.info('archive', archive)
    console.groupEnd()

    let firebase_path = `people/${author}/${type}/`
    if (archive) firebase_path += `${archive}/`
    const folder = await directory(firebase_path)
    folder.items.forEach(item =>
      meta.items.push(/** @type {Id} */ (item.name.split('.')[0]))
    )
    folder.prefixes.forEach(prefix => meta.archive.push(parseInt(prefix.name)))
    await set(path, meta)
    return meta
  }
  return null
}

/**
 * @param {Id} itemid
 * @returns {Promise<Directory | null>}
 */
export const as_directory = async itemid => {
  const path = as_directory_id(itemid)
  const cached = await get(path)
  if (cached) {
    console.info('has directory cached', cached)
    return cached
  }
  let directory = await build_local_directory(itemid)
  if (navigator.onLine)
    try {
      directory = await load_directory_from_network(itemid)
    } catch (e) {
      if (e.code === 'storage/unauthorized') return directory
      throw e
    }

  return directory
}
