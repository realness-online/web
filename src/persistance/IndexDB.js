let openRequest = indexedDB.open("store", 1);

openRequest.onupgradeneeded = ...;
openRequest.onerror = ...;

openRequest.onsuccess = function() {
  let db = openRequest.result;

  db.onversionchange = function() {
    db.close();
    alert("Database is outdated, please reload the page.")
  };

  // ...the db is ready, use it...
};

openRequest.onblocked = function() {
  // there's another open connection to same database
  // and it wasn't closed after db.onversionchange triggered for them
};
