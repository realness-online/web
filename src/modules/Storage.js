import Item from '@/modules/Item'
import * as firebase from 'firebase/app'
import 'firebase/auth'
import 'firebase/storage'

class Storage {
  static hydrate(item_as_string) {
    return document.createRange().createContextualFragment(item_as_string)
  }
  constructor(item_type, selector = `[itemtype="/${item_type}"]`,
    filename = `${item_type}.html`) {
    this.item_type = item_type
    this.selector = selector
    this.filename = filename
    this.metadata = {'contentType': 'text/html'}
  }
  from_storage() {
    let storage_string = localStorage.getItem(this.item_type)
    return Storage.hydrate(storage_string)
  }
  as_list() {
    return Item.get_items(this.from_storage())
  }
  as_object() {
    return Item.get_first_item(this.from_storage())
  }
  save() {
    // save the information to local storage and if appropriate to the server
    let items = document.querySelector(this.selector)
    if (!items) { return false }
    items = items.outerHTML
    localStorage.setItem(this.item_type, items)
    if (['person', 'posts'].includes(this.item_type)) {
      this.persist(items)
    }
    return true
  }
  sync() {
    // check if an 'item_type'.html already exist for this user.
    // If so merge it's content whith what's stored locally
    const url = this.get_server_file_url()
    if (['person', 'posts'].includes(this.item_type) && navigator.onLine) {
      firebase.auth().onAuthStateChanged(user => {
        if (user) {
          const doc_u_path = `/people/${user.phoneNumber}/${this.filename}`
          firebase.storage().ref().child(doc_u_path).getDownloadURL().then((url) => {
            fetch(url).then(response => {
              response.text().then(server_text => {
                const server_as_fragment = Storage.hydrate(server_text)
                let server_as_list = Item.get_items(server_as_fragment)
                console.log(server_as_list)
              })
              let items = [...this.from_storage, ...from_server]
              // console.log(items)
            })
          })
        }
      })
    }
  }

  get_server_file_url() {

  }

  persist(doc_u_ment, doc_u_path) {
    firebase.auth().onAuthStateChanged(user => {
      if (user && navigator.onLine) {
        const file = new File([doc_u_ment], this.filename)
        if (!doc_u_path) {
          doc_u_path = `/people/${user.phoneNumber}/${this.filename}`
        }
        firebase.storage().ref()
          .child(doc_u_path)
          .put(file, this.metadata)
          .catch(console.log.bind(console))
      }
    })
  }
}

export default Storage
export const person_storage = new Storage('person')
export const posts_storage = new Storage('posts', '[itemprop=posts]')
export const activity_storage = new Storage('activity', '[itemprop=activity]')
export const relations = new Storage('relations', '[itemprop=relations]')
export const phonebook = new Storage('phonebook')

//  use this for firebase logging
// .then( snapshot => {
//   console.log('Uploaded', snapshot.totalBytes, 'bytes.')
//   console.log(snapshot.metadata)
//   console.log(snapshot.fullPath)
//   console.log('File available at', snapshot.downloadURL)
// })
