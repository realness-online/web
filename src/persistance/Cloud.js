// https://developers.caffeina.com/object-composition-patterns-in-javascript-4853898bb9d0
import * as firebase from 'firebase/app'
import 'firebase/auth'
import 'firebase/storage'
import { download_url, as_filename } from '@/helpers/itemid'
import { del } from 'idb-keyval'
const networkable = ['person', 'statements', 'posters', 'avatars', 'events']
export const Cloud = (superclass) => class extends superclass {
  async to_network (items) {
    const user = firebase.auth().currentUser
    const path = as_filename(this.id)
    if (user && navigator.onLine) {
      const file = new File([items], path)
      await firebase.storage().ref().child(path).put(file, this.metadata)
    }
  }
  async save (items = document.querySelector(`[itemid="${this.id}"]`)) {
    console.info('Cloud.save()', this.id)
    if (!items) return
    if (super.save) super.save(items)
    if (networkable.includes(this.type)) this.to_network(items.outerHTML)
  }
  async delete () {
    console.info('Cloud.delete()', this.id)
    const user = firebase.auth().currentUser
    if (user && navigator.onLine) {
      del(this.id)
      await firebase.storage().ref().child(as_filename(this.id)).delete()
    }
  }
}
export default Cloud
