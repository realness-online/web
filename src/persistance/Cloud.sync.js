import firebase from 'firebase/app'
import 'firebase/auth'
import 'firebase/storage'
import hash from 'object-hash'
import { get, del, set, keys } from 'idb-keyval'
import { as_filename, as_author, list } from '@/helpers/itemid'
import { Offline } from '@/persistance/Storage'

export const three_minutes = 180000
export const one_hour = 3600000
export const timeouts = []
export const hash_options = { encoding: 'base64', algorithm: 'md5' }
export const does_not_exist = { updated: null, customMetadata: { md5: null } } // Explicitly setting null to indicate that this file doesn't exist

export async function sync_later (id, action) {
  const offline = (await get('sync:offline')) || []
  offline.push({
    id,
    action
  })
  await set('sync:offline', offline)
}
export async function is_stranger (id) {
  const relations = await list(`${localStorage.me}/relations`)
  relations.push({
    id: localStorage.me,
    type: 'person'
  })
  const is_friend = relations.some(relation => {
    if (relation.id === id) return true
    else return false
  })
  return !is_friend
}
export async function sync_offline_actions () {
  if (navigator.onLine) {
    const offline = await get('sync:offline')
    if (!offline) return
    while (offline.length) {
      const item = offline.pop()
      if (item.action === 'save') await new Offline(item.id).save()
      else if (item.action === 'delete') await new Offline(item.id).delete()
      else console.info('weird:unknown-offline-action', item.action, item.id)
    }
    await del('sync:offline')
  }
}
export async function prune () {
  console.log('prune')
  const relations = await list(`${localStorage.me}/relations`)
  const everything = await keys()
  everything.forEach(async itemid => {
    if (!as_author(itemid)) return // items have authors
    if (await is_stranger(as_author(itemid), relations)) await del(itemid)
    if (await itemid.endsWith('/')) await del(itemid) // is a directory
    else {
      const network = await fresh_metadata(itemid)
      if (!network || !network.customMetadata) return null
      const md5 = await local_md5(itemid)
      if (network.customMetadata.md5 !== md5) {
        console.log('outdated', itemid)
        await del(itemid)
      }
    }
  })
}
export async function local_md5 (itemid) { // always checks the network
  const local = await get(itemid)
  if (!local) return null
  return hash(local, hash_options)
}
export async function fresh_metadata (itemid) {
  const index = (await get('sync:index')) || {}
  const path = firebase.storage().ref().child(as_filename(itemid))
  let network
  try {
    // console.info('request:metadata', itemid)
    network = await path.getMetadata()
  } catch (e) {
    if (e.code === 'storage/object-not-found') {
      network = does_not_exist
    } else throw e
  }
  if (!network) throw new Error(`Unable to create metadata for ${itemid}`)
  index[itemid] = network
  await set('sync:index', index)
  return network
}
export function visit_interval () {
  return (Date.now() - one_hour)
}
