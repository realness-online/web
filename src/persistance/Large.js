// https://developers.caffeina.com/object-composition-patterns-in-javascript-4853898bb9d0
import * as firebase from 'firebase/app'
import 'firebase/auth'
import 'firebase/storage'
const Large = (superclass) => class extends superclass {
  async directory () {
    const user = firebase.auth().currentUser
    const storage = firebase.storage().ref()
    if (user && navigator.onLine) {
      const path = `/people/${user.phoneNumber}/${this.type}`
      return storage.child(path).listAll()
    }
  }
}
export default Large
