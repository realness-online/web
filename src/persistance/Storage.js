/** @typedef {import('@/types').Item_Id} Item_Id */

// https://developers.caffeina.com/object-composition-patterns-in-javascript-4853898bb9d0
import Local from '@/persistance/Local'
import Large from '@/persistance/Large'
import Cloud from '@/persistance/Cloud'
import Paged from '@/persistance/Paged'
import { current_user, upload } from '@/use/serverless'
import { as_type, as_filename } from '@/utils/itemid'
import { get } from 'idb-keyval'
import { prepare_upload_html } from '@/utils/upload_processor'

export default class Storage {
  /**
   * @param {Item_Id} itemid
   */
  constructor(itemid) {
    this.id = itemid
    this.type = as_type(itemid)
    this.metadata = { contentType: 'text/html' }
  }
}

/** @extends {Storage} */
export class Poster extends Large(Cloud(Storage)) {}

/** @extends {Storage} */
export class Me extends Cloud(Local(Storage)) {
  constructor() {
    super(localStorage.me)
  }
}

/** @extends {Storage} */
export class Relation extends Local(Storage) {
  constructor() {
    super(`${localStorage.me}/relations`)
  }
}

/** @extends {Storage} */
export class Statement extends Paged(Cloud(Local(Storage))) {
  constructor() {
    super(`${localStorage.me}/statements`)
  }
}

/** @extends {Storage} */
export class Event extends Paged(Cloud(Local(Storage))) {
  constructor() {
    super(`${localStorage.me}/events`)
  }
}

/** @extends {Storage} */
export class Offline extends Cloud(Storage) {
  async save() {
    const outerHTML = await get(this.id)
    if (outerHTML) await super.save({ outerHTML })
  }
}

/** @extends {Storage} */
export class History extends Cloud(Storage) {
  /**
   * @param {Item_Id} itemid
   */
  constructor(itemid) {
    super(itemid)
  }

  async save(items) {
    // on purpose doesn't call super.save
    if (!items) return false
    if (current_user.value && navigator.onLine) {
      const { compressed, metadata } = await prepare_upload_html(items)
      try {
        await upload(as_filename(this.id), compressed, metadata)
      } catch (e) {
        console.error(e)
        return false
      }
      return true
    }
    return false
  }
}
