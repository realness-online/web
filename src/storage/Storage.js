// https://developers.caffeina.com/object-composition-patterns-in-javascript-4853898bb9d0
import Item from '@/modules/item'
import Paged from '@/storage/Paged'
import Large from '@/storage/Large'
import Cloud from '@/storage/Cloud'

class Storage {
  constructor(type,
              selector = `[itemprop="/${type}"]`,
              filename = `${type}/index.html`,
              content_type = 'text/html') {
    this.type = type
    this.selector = selector
    this.filename = filename
    this.metadata = { 'contentType': content_type }
  }
  as_list() {
    return Item.get_items(localStorage.getItem(this.selector))
  }
  as_object() {
    return Item.get_first_item(localStorage.getItem(this.selector))
  }
  async save(items = document.querySelector(this.selector)) {
    if (!items) return
    localStorage.setItem(this.selector, items.outerHTML)
  }
}
class Person extends Cloud(Storage) {
  constructor() {
    super('person', '[itemid="person.html"]', 'person.html')
  }
}
class Post extends Paged(Cloud(Storage)) {
  constructor() {
    super('posts')
  }
}
class SVG extends Large(Cloud(Storage)) {}

export default Storage
export const relations_storage = new Storage('relations')
export const person_storage = new Person()
export const posts_storage = new Posts()
export const avatars_storage = new SVG('avatars')
export const posters_storage = new SVG('posters')
