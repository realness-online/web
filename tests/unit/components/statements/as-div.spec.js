import { shallowMount } from '@vue/test-utils'
import flushPromises from 'flush-promises'
import as_statement from '@/components/statements/as-div'
import get_item from '@/modules/item'
import { Statements } from '@/persistance/Storage'
const statements_as_html = require('fs').readFileSync('./tests/unit/html/statements.html', 'utf8')
describe('@/components/statements/as-div.vue', () => {
  let wrapper
  const statements = get_item(statements_as_html)
  beforeEach(() => {
    wrapper = shallowMount(as_statement, { propsData: { statement: statements.statements[0] } })
  })
  describe('rendering:', () => {
    it('Render a statement', async () => {
      await flushPromises()
      expect(wrapper.element).toMatchSnapshot()
      wrapper.destroy()
    })
  })
  describe('methods', () => {
    describe('#save', () => {
      it('Saves a statement', async () => {
        wrapper = shallowMount(as_statement, {
           propsData: {
             editable: true,
             statement: statements.statements[0]
           }
        })
        jest.spyOn(Statements.prototype, 'save').mockImplementation(_ => {
          return jest.fn(() => Promise.resolve())
        })
        wrapper.vm.$refs.editable.textContent = 'changed'
        await wrapper.vm.save()
        expect(Statements.prototype.save).toBeCalled()
      })
      it('Doe nothing if the statement is the same', async () => {
        wrapper = shallowMount(as_statement, {
           propsData: {
             editable: true,
             statement: statements.statements[0]
           }
        })
        jest.spyOn(Statements.prototype, 'save').mockImplementation(_ => {
          return jest.fn(() => Promise.resolve())
        })
        await wrapper.vm.save()
        expect(Statements.prototype.save).not.toBeCalled()
      })
    })
    describe('#focused', () => {
      it('Emits an event', () => {
        wrapper.vm.focused()
        expect(wrapper.emitted('focused')).toBeTruthy()
      })
    })
  })
})
