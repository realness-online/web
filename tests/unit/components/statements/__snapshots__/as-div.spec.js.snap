// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@/components/statements/as-div.vue rendering: Render a statement 1`] = `
<div
  itemid="/+16282281824/statements/1569909292018"
  itemscope="itemscope"
>
  <p
    itemprop="statement"
  >
    Mark Zuckerberg is just feeding his AI vanity
  </p>
   
  <!---->
   
  <!---->
</div>
`;
